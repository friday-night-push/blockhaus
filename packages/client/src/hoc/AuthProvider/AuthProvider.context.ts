import type { Dispatch, SetStateAction } from 'react';
import { createContext } from 'react';

import type { Nullable } from 'src/shared/types/global';
import type { TUser } from 'src/shared/types/user';

export interface AuthContextType {
  user: Nullable<TUser>;
  setUser?: Dispatch<SetStateAction<Nullable<TUser>>>;
  userIsLoading?: boolean;
  setUserIsLoading?: Dispatch<SetStateAction<boolean>>;
}

export const AuthContext = createContext<AuthContextType>({
  user: null,
  setUser: undefined,
  userIsLoading: false,
  setUserIsLoading: undefined,
});
