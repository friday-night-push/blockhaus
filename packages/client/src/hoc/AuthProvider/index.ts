export * from './AuthProvider.context';

export * from './AuthProvider';
