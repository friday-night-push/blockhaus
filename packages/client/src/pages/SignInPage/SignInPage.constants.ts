import * as Yup from 'yup';

import type { InputProps } from 'src/components/atoms/Input';

export const USER_DATA_MOCK = {
  login: 'l586',
  password: '123!!321',
};

export const inputs: InputProps[] = [
  { name: 'login', label: 'Login', type: 'text' },
  { name: 'password', label: 'Password', type: 'password' },
];

export const validationSchema = Yup.object({
  login: Yup.string()
    .max(15, 'Must be 15 characters or less')
    .required('Required'),
  password: Yup.string()
    .min(8, 'Password should be of minimum 8 characters length')
    .required('Password is required'),
});
