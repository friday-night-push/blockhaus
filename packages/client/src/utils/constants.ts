export const HOST = 'http://friday-night-push-blockhau-40.ya-praktikum.tech';
export const SERVER_HOST_URL = `${HOST}:3001`;
export const API_URL = `${SERVER_HOST_URL}/api/v1`;
export const YANDEX_URL = `${API_URL}/proxy`;

// OAuth
export const REDIRECT_URL = `${HOST}:3000`;

export const RESOURCE_URL = YANDEX_URL + '/resources';

export enum PAGE_ROUTES {
  MENU = '/',
  SIGN_IN = '/sign-in',
  SIGN_UP = '/sign-up',
  CHANGE_PASSWORD = '/change-password',
  RECOVER_PASSWORD = '/recover-password',
  PROFILE = '/profile',
  GAME = '/game',
  GAME_PAUSE = '/game-pause',
  GAME_RESULT = '/game-result',
  GAME_SETUP = '/game-setup',
  GAME_DIFFICULT = '/game-difficult',
  LEADER_BOARD = '/leader-board',
  ERROR_BOUNDARY = '/error-boundary',
  FORUM = '/forum',
  FORUM_TOPIC = '/forum/:topicId/:page?',
}
