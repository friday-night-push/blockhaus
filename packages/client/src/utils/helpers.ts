// Отдельный класс исключений для ESLint

export default class Helpers {
  public static Log(level: string, ...args: any): any {
    switch (level) {
      case 'INFO':
        // eslint-disable-next-line no-console
        console.info(...args);
        break;
      case 'LOG':
        // eslint-disable-next-line no-console
        console.log(...args);
        break;
      case 'ERROR':
        // eslint-disable-next-line no-console
        console.error(...args);
        break;
      default:
        break;
    }
  }
}
