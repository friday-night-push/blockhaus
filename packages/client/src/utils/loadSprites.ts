export function loadSprites(
  spritesPath: Record<string, string> = {}
): Promise<Record<string, HTMLImageElement>> {
  return Promise.all(
    Object.values(spritesPath).map(path => loadImage(path))
  ).then(sprites => {
    const keys = Object.keys(spritesPath);
    const map = keys.map((key, index) => [key, sprites[index]]);
    return Object.fromEntries(map);
  }) as Promise<Record<string, HTMLImageElement>>;
}

function loadImage(src: string): Promise<HTMLImageElement> {
  return new Promise<HTMLImageElement>((resolve, reject) => {
    const img = new Image();
    img.onload = () => resolve(img);
    img.onerror = reject;
    img.src = src;
  });
}
