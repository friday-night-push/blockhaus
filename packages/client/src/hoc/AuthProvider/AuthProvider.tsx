import type { ReactNode } from 'react';
import { useContext, useEffect, useState } from 'react';

import AuthAPI from 'src/services/api/auth-api';
import type { Nullable } from 'src/shared/types/global';
import type { TUser } from 'src/shared/types/user';
import Helpers from 'src/utils/helpers';

import type { AuthContextType } from './AuthProvider.context';
import { AuthContext } from './AuthProvider.context';

export const authAPI = new AuthAPI();

export const AuthProvider = ({ children }: { children: ReactNode }) => {
  const [user, setUser] = useState<Nullable<TUser>>(null);
  const [userIsLoading, setUserIsLoading] = useState(true);

  const isAuthenticated = async () => {
    setUserIsLoading(true);

    try {
      const isAuth = localStorage.getItem('isAuth');
      if (isAuth !== null) await authAPI.getuser(updateUser, errorHandler);
    } catch (err) {
      errorHandler(err);
    } finally {
      setUserIsLoading(false);
    }
  };

  const updateUser = (user: TUser) => {
    localStorage.setItem('isAuth', 'true');
    setUser(user);
  };

  const errorHandler = (err: Error) => {
    Helpers.Log('ERROR', err);
  };

  useEffect(() => {
    isAuthenticated();
  }, []);

  return (
    <AuthContext.Provider
      value={{ user, userIsLoading, setUser, setUserIsLoading }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = (): AuthContextType => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};
