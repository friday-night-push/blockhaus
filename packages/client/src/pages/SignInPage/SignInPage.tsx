import React, { useContext, useState } from 'react';

import { Text } from '@gravity-ui/uikit';
import { Navigate, useNavigate } from 'react-router-dom';

import { Button, Container, Form, Logo, Page } from 'src/components';
import { AuthContext } from 'src/hoc';
import AuthAPI from 'src/services/api/auth-api';
import { TSignInRequest, TUser } from 'src/shared/types/user';
import { PAGE_ROUTES } from 'src/utils/constants';
import Helpers from 'src/utils/helpers';

import { inputs, validationSchema } from './SignInPage.constants';

export const authAPI = new AuthAPI();

export const SignInPage: React.FC = () => {
  const { user, setUser, userIsLoading } = useContext(AuthContext);
  const [error, setError] = useState<string | undefined>(undefined);

  const navigate = useNavigate();

  const auth = async (formData: TSignInRequest) => {
    await authAPI.signin(formData, isOk, errorHandler);
  };

  const isOk = () => {
    authAPI.getuser(updUserData, errorHandler);
  };

  const updUserData = (user: TUser) => {
    if (setUser) {
      setUser(user);
    }
  };

  const errorHandler = (err: unknown) => {
    setError(String(err));
    Helpers.Log('ERROR', err);
  };

  const goToSignUp = () => {
    navigate(PAGE_ROUTES.SIGN_UP);
  };

  const goBack = () => {
    navigate(-1);
  };

  return (
    <Page>
      {!user?.id && !userIsLoading ? (
        <>
          <Logo size="sm" />
          <Container direction="column" alignItems="center">
            <Text variant="display-2">Sign in</Text>
            <Text variant="subheader-2" style={{ textAlign: 'center' }}>
              Back again? Just sign in to keep your results showing up on the
              leaderboards
            </Text>
          </Container>
          <Form
            inputs={inputs}
            validationSchema={validationSchema}
            onSubmit={auth}
            errorMessage={error}
          />
          <Button view="flat" onClick={goToSignUp}>
            First time here? Sign up
          </Button>
          <Button view="flat" onClick={goBack}>
            Back
          </Button>
        </>
      ) : (
        <Navigate to={PAGE_ROUTES.MENU} />
      )}
    </Page>
  );
};
