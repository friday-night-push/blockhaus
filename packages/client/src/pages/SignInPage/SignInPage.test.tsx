import userEvent from '@testing-library/user-event';

import { PAGE_ROUTES } from 'src/utils/constants';
import { render, screen } from 'src/utils/tests';

import { mockedUseNavigate, mockedUser } from 'src/utils/tests/mocks';

import { SignInPage } from './SignInPage';
import { USER_DATA_MOCK } from './SignInPage.constants';
import { authAPI } from '../../hoc/AuthProvider';

describe('SignInPage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    const { container } = render(<SignInPage />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('navigates to menu if user is authenticated', async () => {
    render(<SignInPage />, {
      authState: {
        user: mockedUser,
        userIsLoading: false,
      },
    });

    expect(screen.queryByText('Sign in')).not.toBeInTheDocument();
  });

  it('navigates back on back button click', async () => {
    render(<SignInPage />);

    const user = userEvent.setup();
    await user.click(screen.getByRole('button', { name: 'Back' }));
    expect(mockedUseNavigate).toHaveBeenCalledWith(-1);
  });

  it('navigates to sign up on sign up button click', async () => {
    render(<SignInPage />);

    const user = userEvent.setup();
    await user.click(
      screen.getByRole('button', { name: 'First time here? Sign up' })
    );
    expect(mockedUseNavigate).toHaveBeenCalledWith(PAGE_ROUTES.SIGN_UP);
  });

  it('displays an error message on failed sign in', async () => {
    jest.spyOn(authAPI, 'signin').mockImplementation((_, __, errorCb) => {
      errorCb('Invalid credentials');
      return Promise.resolve();
    });

    render(<SignInPage />);

    const user = userEvent.setup();

    await user.type(screen.getByLabelText(/login/i), 'wrong_login');
    await user.type(screen.getByLabelText(/password/i), 'wrong_password');
    await user.click(screen.getByRole('button', { name: /submit/i }));

    expect(await screen.findByText(/Invalid credentials/i)).toBeInTheDocument();
  });

  it('calls signin with correct data', async () => {
    jest.spyOn(authAPI, 'signin').mockImplementation((data, cb) => {
      cb(new Response(undefined, { status: 200 }));
      return Promise.resolve();
    });

    render(<SignInPage />);

    const user = userEvent.setup();

    await user.type(screen.getByLabelText(/login/i), USER_DATA_MOCK.login);
    await user.type(
      screen.getByLabelText(/password/i),
      USER_DATA_MOCK.password
    );
    await user.click(screen.getByRole('button', { name: /submit/i }));

    expect(authAPI.signin).toHaveBeenCalledWith(
      { login: USER_DATA_MOCK.login, password: USER_DATA_MOCK.password },
      expect.any(Function),
      expect.any(Function)
    );
  });
});
