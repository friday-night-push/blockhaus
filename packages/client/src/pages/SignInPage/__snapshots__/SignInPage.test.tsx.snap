// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInPage matches snapshot 1`] = `
<section
  class="g-box g-flex"
  style="width: auto; min-height: 100vh; column-gap: 12px; row-gap: 12px; justify-content: center;"
>
  <section
    class="g-box g-flex"
    style="width: 100%; min-height: 100vh; max-width: 380px; flex-direction: column; column-gap: 32px; row-gap: 32px; align-items: center; justify-content: center;"
  >
    <img
      alt="Blockhaus"
      style="display: flex; width: 240px;"
    />
    <section
      class="g-box g-flex"
      style="flex-direction: column; column-gap: 12px; row-gap: 12px; align-items: center;"
    >
      <span
        class="g-text g-text_variant_display-2"
      >
        Sign in
      </span>
      <span
        class="g-text g-text_variant_subheader-2"
        style="text-align: center;"
      >
        Back again? Just sign in to keep your results showing up on the leaderboards
      </span>
      <form
        action="#"
        style="width: 100%;"
      >
        <section
          class="g-box g-flex"
          style="width: 100%; flex-direction: column; column-gap: 16px; row-gap: 16px; align-items: center;"
        >
          <span
            class="g-text-input g-text-input_view_normal g-text-input_size_xl g-text-input_pin_brick-brick"
          >
            <span
              class="g-text-input__content"
            >
              <label
                class="g-text-input__label"
                for="g-:r0:"
                style="max-width: calc(50% - 0px);"
                title="Login"
              >
                Login
              </label>
              <input
                class="g-text-input__control g-text-input__control_type_input"
                id="g-:r0:"
                name="login"
                type="text"
                value=""
              />
            </span>
          </span>
          <span
            class="g-text-input g-text-input_view_normal g-text-input_size_xl g-text-input_pin_brick-brick"
          >
            <span
              class="g-text-input__content"
            >
              <label
                class="g-text-input__label"
                for="g-:r3:"
                style="max-width: calc(50% - 0px);"
                title="Password"
              >
                Password
              </label>
              <input
                class="g-text-input__control g-text-input__control_type_input"
                id="g-:r3:"
                name="password"
                type="password"
                value=""
              />
            </span>
          </span>
          <section
            class="g-box g-flex"
            style="flex-direction: row; column-gap: 8px; row-gap: 8px;"
          >
            <button
              class="g-button g-button_view_action g-button_size_xl g-button_pin_brick-brick g-button_width_max"
              type="submit"
            >
              <span
                class="g-button__text"
              >
                Submit
              </span>
            </button>
          </section>
        </section>
      </form>
      <a
        aria-disabled="false"
        class="g-button g-button_view_flat g-button_size_xl g-button_pin_brick-brick"
        href="/sign-up"
      >
        <span
          class="g-button__text"
        >
          First time here? Sign up
        </span>
      </a>
      <button
        class="g-button g-button_view_flat g-button_size_xl g-button_pin_brick-brick"
        type="button"
      >
        <span
          class="g-button__text"
        >
          Go Back
        </span>
      </button>
    </section>
  </section>
</section>
`;
